File: invite/invite.go
package invite

import (
	"crypto/ed25519"
	"encoding/base64"
	"fmt"
	"net/url"

	"github.com/corpo/qntm/crypto"
	"github.com/corpo/qntm/identity"
	"github.com/corpo/qntm/pkg/cbor"
	"github.com/corpo/qntm/pkg/types"
)

// Manager handles invite creation, parsing, and key schedule derivation
type Manager struct {
	suite       *crypto.QSP1Suite
	identityMgr *identity.Manager
}

// NewManager creates a new invite manager
func NewManager() *Manager {
	return &Manager{
		suite:       crypto.NewQSP1Suite(),
		identityMgr: identity.NewManager(),
	}
}

// CreateInvite creates a new conversation invite
func (m *Manager) CreateInvite(inviterIdentity *types.Identity, convType types.ConversationType) (*types.InvitePayload, error) {
	if err := m.identityMgr.ValidateIdentity(inviterIdentity); err != nil {
		return nil, fmt.Errorf("invalid inviter identity: %w", err)
	}
	
	// Generate conversation ID
	convID, err := m.identityMgr.GenerateConversationID()
	if err != nil {
		return nil, fmt.Errorf("failed to generate conversation ID: %w", err)
	}
	
	// Generate invite secret (32 bytes high entropy)
	inviteSecret, err := m.identityMgr.GenerateRandomBytes(32)
	if err != nil {
		return nil, fmt.Errorf("failed to generate invite secret: %w", err)
	}
	
	// Generate invite salt (16-32 bytes, using 32 for extra security)
	inviteSalt, err := m.identityMgr.GenerateRandomBytes(32)
	if err != nil {
		return nil, fmt.Errorf("failed to generate invite salt: %w", err)
	}
	
	invite := &types.InvitePayload{
		Version:      types.ProtocolVersion,
		Suite:        types.DefaultSuite,
		Type:         string(convType),
		ConvID:       convID,
		InviterIKPK:  inviterIdentity.PublicKey,
		InviteSalt:   inviteSalt,
		InviteSecret: inviteSecret,
	}
	
	return invite, nil
}

// SerializeInvite serializes an invite to canonical CBOR
func (m *Manager) SerializeInvite(invite *types.InvitePayload) ([]byte, error) {
	return cbor.MarshalCanonical(invite)
}

// DeserializeInvite deserializes an invite from canonical CBOR
func (m *Manager) DeserializeInvite(data []byte) (*types.InvitePayload, error) {
	var invite types.InvitePayload
	if err := cbor.UnmarshalCanonical(data, &invite); err != nil {
		return nil, fmt.Errorf("failed to unmarshal invite: %w", err)
	}
	
	// Validate the invite
	if err := m.ValidateInvite(&invite); err != nil {
		return nil, fmt.Errorf("invalid invite: %w", err)
	}
	
	return &invite, nil
}

// ValidateInvite validates the structure and content of an invite
func (m *Manager) ValidateInvite(invite *types.InvitePayload) error {
	if invite == nil {
		return fmt.Errorf("invite is nil")
	}
	
	// Check version
	if invite.Version != types.ProtocolVersion {
		return fmt.Errorf("unsupported protocol version: %d", invite.Version)
	}
	
	// Check suite
	if invite.Suite != types.DefaultSuite {
		return fmt.Errorf("unsupported crypto suite: %s", invite.Suite)
	}
	
	// Check conversation type
	if invite.Type != string(types.ConversationTypeDirect) && invite.Type != string(types.ConversationTypeGroup) {
		return fmt.Errorf("invalid conversation type: %s", invite.Type)
	}
	
	// Check public key length
	if len(invite.InviterIKPK) != ed25519.PublicKeySize {
		return fmt.Errorf("invalid inviter public key length: %d", len(invite.InviterIKPK))
	}
	
	// Check invite salt length (should be 16-32 bytes)
	if len(invite.InviteSalt) < 16 || len(invite.InviteSalt) > 32 {
		return fmt.Errorf("invalid invite salt length: %d", len(invite.InviteSalt))
	}
	
	// Check invite secret length (must be exactly 32 bytes for security)
	if len(invite.InviteSecret) != 32 {
		return fmt.Errorf("invalid invite secret length: %d", len(invite.InviteSecret))
	}
	
	return nil
}

// InviteToToken encodes an invite as a base64url token string (no URL wrapping).
func (m *Manager) InviteToToken(invite *types.InvitePayload) (string, error) {
	inviteData, err := m.SerializeInvite(invite)
	if err != nil {
		return "", fmt.Errorf("failed to serialize invite: %w", err)
	}
	return base64.RawURLEncoding.EncodeToString(inviteData), nil
}

// InviteToURL encodes an invite as a URL with the invite data in the fragment
func (m *Manager) InviteToURL(invite *types.InvitePayload, baseURL string) (string, error) {
	// Serialize the invite
	inviteData, err := m.SerializeInvite(invite)
	if err != nil {
		return "", fmt.Errorf("failed to serialize invite: %w", err)
	}
	
	// Base64 encode for URL safety
	inviteB64 := base64.RawURLEncoding.EncodeToString(inviteData)
	
	// Create URL with invite in fragment to reduce server logging
	parsedURL, err := url.Parse(baseURL)
	if err != nil {
		return "", fmt.Errorf("invalid base URL: %w", err)
	}
	
	parsedURL.Fragment = inviteB64
	
	return parsedURL.String(), nil
}

// InviteFromURL extracts an invite from a URL fragment
func (m *Manager) InviteFromURL(inviteURL string) (*types.InvitePayload, error) {
	parsedURL, err := url.Parse(inviteURL)
	if err != nil {
		return nil, fmt.Errorf("invalid invite URL: %w", err)
	}
	
	if parsedURL.Fragment == "" {
		return nil, fmt.Errorf("no invite data in URL fragment")
	}
	
	// Decode base64
	inviteData, err := base64.RawURLEncoding.DecodeString(parsedURL.Fragment)
	if err != nil {
		return nil, fmt.Errorf("failed to decode invite data: %w", err)
	}
	
	// Deserialize the invite
	return m.DeserializeInvite(inviteData)
}

// DeriveConversationKeys derives conversation keys from an invite
func (m *Manager) DeriveConversationKeys(invite *types.InvitePayload) (*types.ConversationKeys, error) {
	if err := m.ValidateInvite(invite); err != nil {
		return nil, fmt.Errorf("invalid invite: %w", err)
	}
	
	convIDBytes := invite.ConvID[:]
	
	// Derive root key
	rootKey, err := m.suite.DeriveRootKey(invite.InviteSecret, invite.InviteSalt, convIDBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to derive root key: %w", err)
	}
	
	// Derive conversation keys
	aeadKey, nonceKey, err := m.suite.DeriveConversationKeys(rootKey, convIDBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to derive conversation keys: %w", err)
	}
	
	keys := &types.ConversationKeys{
		Root:     rootKey,
		AEADKey:  aeadKey,
		NonceKey: nonceKey,
	}
	
	return keys, nil
}

// CreateConversation creates a conversation from an invite and keys
func (m *Manager) CreateConversation(invite *types.InvitePayload, keys *types.ConversationKeys) (*types.Conversation, error) {
	if err := m.ValidateInvite(invite); err != nil {
		return nil, fmt.Errorf("invalid invite: %w", err)
	}
	
	// Initial participants include the inviter
	inviterKeyID := m.identityMgr.KeyIDFromPublicKey(invite.InviterIKPK)
	participants := []types.KeyID{inviterKeyID}
	
	conversation := &types.Conversation{
		ID:           invite.ConvID,
		Type:         types.ConversationType(invite.Type),
		Keys:         *keys,
		Participants: participants,
	}
	
	return conversation, nil
}

// AddParticipant adds a participant to a conversation
func (m *Manager) AddParticipant(conv *types.Conversation, pubkey ed25519.PublicKey) {
	keyID := m.identityMgr.KeyIDFromPublicKey(pubkey)
	
	// Check if participant already exists
	for _, existing := range conv.Participants {
		if existing == keyID {
			return // Already a participant
		}
	}
	
	conv.Participants = append(conv.Participants, keyID)
}

// IsParticipant checks if a public key is a participant in the conversation
func (m *Manager) IsParticipant(conv *types.Conversation, pubkey ed25519.PublicKey) bool {
	keyID := m.identityMgr.KeyIDFromPublicKey(pubkey)
	
	for _, participant := range conv.Participants {
		if participant == keyID {
			return true
		}
	}
	
	return false
}

// GenerateInviteSecret generates a high-entropy 32-byte invite secret
func (m *Manager) GenerateInviteSecret() ([]byte, error) {
	return m.identityMgr.GenerateRandomBytes(32)
}

// GenerateInviteSalt generates a random invite salt
func (m *Manager) GenerateInviteSalt() ([]byte, error) {
	return m.identityMgr.GenerateRandomBytes(32)
}

File: python-dist/src/qntm/invite.py
"""Invite creation, parsing, and conversation bootstrapping."""

import os

from .cbor import marshal_canonical, unmarshal
from .constants import DEFAULT_SUITE, PROTOCOL_VERSION
from .crypto import QSP1Suite
from .identity import (
    base64url_decode,
    base64url_encode,
    generate_conversation_id,
    key_id_from_public_key,
    validate_identity,
)

_suite = QSP1Suite()


def create_invite(inviter_identity: dict, conv_type: str = "direct") -> dict:
    """Create a new invite payload."""
    validate_identity(inviter_identity)

    conv_id = generate_conversation_id()
    invite_secret = os.urandom(32)
    invite_salt = os.urandom(32)

    return {
        "v": PROTOCOL_VERSION,
        "suite": DEFAULT_SUITE,
        "type": conv_type,
        "conv_id": conv_id,
        "inviter_ik_pk": inviter_identity["publicKey"],
        "invite_salt": invite_salt,
        "invite_secret": invite_secret,
    }


def serialize_invite(invite: dict) -> bytes:
    return marshal_canonical(invite)


def deserialize_invite(data: bytes) -> dict:
    invite = unmarshal(data)
    validate_invite(invite)
    return invite


def validate_invite(invite: dict) -> None:
    if invite["v"] != PROTOCOL_VERSION:
        raise ValueError(f"unsupported protocol version: {invite['v']}")
    if invite["suite"] != DEFAULT_SUITE:
        raise ValueError(f"unsupported crypto suite: {invite['suite']}")
    if invite["type"] not in ("direct", "group"):
        raise ValueError(f"invalid conversation type: {invite['type']}")
    inviter_pk = invite["inviter_ik_pk"]
    if isinstance(inviter_pk, memoryview):
        inviter_pk = bytes(inviter_pk)
    if len(inviter_pk) != 32:
        raise ValueError(f"invalid inviter public key length: {len(inviter_pk)}")
    invite_salt = invite["invite_salt"]
    if isinstance(invite_salt, memoryview):
        invite_salt = bytes(invite_salt)
    salt_len = len(invite_salt)
    if salt_len < 16 or salt_len > 32:
        raise ValueError(f"invalid invite salt length: {salt_len}")
    invite_secret = invite["invite_secret"]
    if isinstance(invite_secret, memoryview):
        invite_secret = bytes(invite_secret)
    if len(invite_secret) != 32:
        raise ValueError(f"invalid invite secret length: {len(invite_secret)}")


def invite_to_token(invite: dict) -> str:
    data = serialize_invite(invite)
    return base64url_encode(data)


def invite_from_url(invite_url: str) -> dict:
    """Parse an invite from a URL fragment or bare token."""
    fragment = invite_url
    if "://" in invite_url:
        # Extract fragment from URL
        parts = invite_url.split("#", 1)
        if len(parts) < 2 or not parts[1]:
            raise ValueError("no invite data in URL fragment")
        fragment = parts[1]

    data = base64url_decode(fragment)
    return deserialize_invite(data)


def derive_conversation_keys(invite: dict) -> dict:
    """Derive root, AEAD, and nonce keys from invite."""
    validate_invite(invite)

    invite_secret = bytes(invite["invite_secret"])
    invite_salt = bytes(invite["invite_salt"])
    conv_id = bytes(invite["conv_id"])

    root_key = _suite.derive_root_key(invite_secret, invite_salt, conv_id)
    aead_key, nonce_key = _suite.derive_conversation_keys(root_key, conv_id)

    return {
        "root": root_key,
        "aeadKey": aead_key,
        "nonceKey": nonce_key,
    }


def create_conversation(invite: dict, keys: dict) -> dict:
    """Create a conversation object from invite + derived keys."""
    validate_invite(invite)

    inviter_key_id = key_id_from_public_key(bytes(invite["inviter_ik_pk"]))

    return {
        "id": bytes(invite["conv_id"]),
        "type": invite["type"],
        "keys": keys,
        "participants": [inviter_key_id],
        "currentEpoch": 0,
    }


def add_participant(conv: dict, public_key: bytes) -> None:
    """Add a participant to a conversation."""
    key_id = key_id_from_public_key(public_key)
    for existing in conv["participants"]:
        if existing == key_id:
            return
    conv["participants"].append(key_id)
